src/js/components/AppHealthBarWithTooltipComponent.jsx

```diff
@@ -21,15 +21,20 @@ var AppHealthBarWithTooltipComponent = React.createClass({
     };
   },
 
-  shouldComponentUpdate: function (nextProps) {
-    var shouldUpdate =
+  shouldComponentUpdate: function (nextProps, nextState) {
+    // Ensure that the health bar does not update before the tipContent was
+    // updated (`setState` is not guaranteed to complete before `render`)
+    if (this.state.tipContent !== nextState.tipContent) {
+      return true;
+    }
+    var healthWasUpdated =
       !Util.compareArrays(this.props.model.health, nextProps.model.health);
-    if (shouldUpdate) {
+    if (healthWasUpdated) {
       this.setState({
         tipContent: this.getAppHealthBreakdown(nextProps.model)
       });
     }
-    return shouldUpdate;
+    return false;
```

Comment:
You're right, it's not possible, and even if we aren't sure about the order of execution (`return false` might actually happen before `setState` has actually updated the state) we perform state checks at the beginning of the `shouldComponentUpdate`. Sorry for the intrusion!